app/src/main/java/nextstep/github/core/model/OrganizationRepository.kt

```diff
@@ -0,0 +1,23 @@
+package nextstep.github.core.model
+
+sealed interface OrganizationRepository {
+    val fullName: String
+    val description: String
+    val stars: Int
+
+    data class Hot(
+        override val fullName: String,
+        override val description: String,
+        override val stars: Int,
+    ) : OrganizationRepository
+
+    data class Normal(
+        override val fullName: String,
+        override val description: String,
+        override val stars: Int,
+    ) : OrganizationRepository
+
+    companion object {
+        const val HOT_STARS = 50
+    }
+}
```

Comment:
다음과 같이 구현하는 것과 어떤 차이가 있을까요?
```suggestion
data class OrganizationRepository(
    val fullName: String,
    val description: String,
    val stars: Int,
    val isHot: Boolean,
)
```

Reply:
추가적인 값이 추가 되었을 때 수현님이 적어주신 코드가 훨씬 유리 합니다. 제가 작성한 코드는 Hot이나 Normal에서만 별도로 사용할 값들이 생길 때만 유효하지 동일한 값들이 추가 되었을 때에는 코드 수만 늘어나서 의미없는 구조체가 되버리죠.

이 객체에 역할에 대해서 좀 더 생각해보다가 수현님이 작성해주신 코드로 변경했습니다. 추가적으로 해당 Entity이 역할에 대해서 고민하다가 필요 없다 결론이 나서 삭제했습니다. 
https://github.com/next-step/android-github-compose/pull/13/commits/f97d9cdde6159a34df8e314f2c34ec66dd78eb0c

그 이유는 ......좀 길지만 제 생각을 적어봤습니다.

1. `OrganizationRepository` Entity는 확장성도 없고 유연하지 못한 오히려 제한적 요소를 가미시키는 객체이다.
- 말 그대로 위에서 애기 했듯이 어떤 값이 추가 될 때마다 어디에 넣어야 할지 고민을 하게 되고 코드가 중복되는 등의 불편한 사항들이 많이 질 것으로 보입니다.

2. `OrganizationRepository`는 선택 사항에 충족시키고자 억지로 만든 객체였다. 
- 선택 사항에서 (optional인 )domain을 추가 해서 `별 50개 이상의 HOT` 기준을 결정하는 것을 고민 하다 보니 sealed interface인  `OranizationRepository`를 만들어서 `Hot`, `Normal` 도메인을 만들어서 관리 하려고 했던 의도였지만, `Hot`을 결정하는게 도메인의 역할일까 라는 고민이 들었습니다.

3. `GetOrganizationRepositoryUseCase`는 의미 있는 도메인 UseCase가 (개인적으로)아니다.
- 1번의 연장선으로 선택 사항에 대한 구현을 충족시키고자 구현을 하게 되었습니다. 하지만 `별 50개 이상의 HOT` 기준을 정하는게 정말 UseCase에서 필요한 비즈니스 로직일까? 라는 고민을 해보았고 결론은 아니다로 났습니다. 그 이유는 `Hot` 라는 상태는 Entity의 star 개수에 의해서 정해지는 것이 기 때문에 오히려 `Entity` 즉, `model`의 비즈니스 로직으로 생각했습니다. 

이러한 이유로 기존에 작성한 `RespositoryEntity`에 `isHot`을 추가하고 `OrganizationRepository`와 관련된 코드는 모두 삭제 한 상태입니다.

너무 아키텍처 구조에 맞게끔 독립적으로 각자 역할과 의미를 부여해서 작성해보려다가 잘못된 방향으로 흘러간 것 같아요.  개인적으로는 제 실력도 부족하고 이러한 사항이 너무 지나치게 클린 아키텍처를 지키려도 보니 오버 엔지니어링으로 이어지지 않았나 생각이 드네요. 사견이지만 이러한 이유 때문에 클린 아키텍처, 객체 지향...이런 것들이 저는 좋아하지 않고 불편한 것들이 좀 많이 있다고 개인적으로 생각을 많이해요. 그걸 지키기 위해서 구조를 독립시키는게 아니라 감옥을 만들어서 감옥에 가둔다고나 할까나....또 그 규칙을 지키기 위해서 의미 없는 보일러플레이트성 코드도 양산하는 것 같다는 느낌도 들구요. 

이러한 이유들 때문에 수현님도 잘 아시는 분인 플러터 개발자 분께서 파이브 라인스 책을 읽고 나서 한 말이 저는 정말 공감이 가더라구요.
> 파이브 라인스 저자가 책에서 하고 싶은 말은 객체 지향, 클린 코드, 클린 아키텍처...이런거 다 필요 없고 함수만 잘 정리해서 선언하더라도 충분히 유지보수 가능한 코드가 된다. 라고 말을 하고 싶은 것 같다.

Reply:
민재님께서 도메인 레이어에 어떤 내용이 들어가야 할지 이미 충분히 합리적으로 고민을 해주신 것 같아 이 구조에 대해서는 더 피드백 드리지 않아도 될 것 같아요!

> 너무 아키텍처 구조에 맞게끔 독립적으로 각자 역할과 의미를 부여해서 작성해보려다가 잘못된 방향으로 흘러간 것 같아요. 개인적으로는 제 실력도 부족하고 이러한 사항이 너무 지나치게 클린 아키텍처를 지키려도 보니 오버 엔지니어링으로 이어지지 않았나 생각이 드네요. 사견이지만 이러한 이유 때문에 클린 아키텍처, 객체 지향...이런 것들이 저는 좋아하지 않고 불편한 것들이 좀 많이 있다고 개인적으로 생각을 많이해요. 그걸 지키기 위해서 구조를 독립시키는게 아니라 감옥을 만들어서 감옥에 가둔다고나 할까나....또 그 규칙을 지키기 위해서 의미 없는 보일러플레이트성 코드도 양산하는 것 같다는 느낌도 들구요.
마지막으로 뜬금 없지만 수현님도 잘 아시는 분인 플러터 개발자 분께서 파이브 라인스 책을 읽고 나서 한 말이 저는 정말 공감이 가더라구요.
> > 파이브 라인스 저자가 책에서 하고 싶은 말은 객체 지향, 클린 코드, 클린 아키텍처...이런거 다 필요 없고 함수만 잘 정리해서 선언하더라도 충분히 유지보수 가능한 코드가 된다. 라고 말을 하고 싶은 것 같다.



마지막에 재밌는 주제를 남겨주셔서 제 의견도 남깁니다!

클린 아키텍처(아마 이 맥락에서는 권장 앱 아키텍처를 말씀하고자 하신 것이겠죠...?)나 객체지향을 지나치게 강조하다 보면 코드가 복잡해지고 오버 엔지니어링으로 이어질 수 있다는 점, 그리고 의미 없는 보일러플레이트 코드를 양산할 수 있다는 점도 충분히 공감됩니다.

하지만 객체지향과 권장 앱 아키텍처의 본질을 제대로 이해한다면, 그러한 문제점들을 최소화하고 실제로 코드의 유지보수성과 확장성을 높일 수 있습니다. 핵심은 "클래스"가 아니라 "메시지"에 있다는 것을 잊지 말아야 합니다. 객체지향의 핵심은 클래스를 정의하고 상속하는 것이 아니라, 객체 간의 협력을 통해 메시지를 주고받으며 문제를 해결하는 데 있습니다. 
([창시자 앨런 케이가 말하는, 객체 지향 프로그래밍의 본질](https://velog.io/@eddy_song/alan-kay-OOP))

아키텍처 또한 마찬가지입니다. 권장 앱 아키텍처는 단순히 규칙을 강요하는 것이 아니라, 코드의 의존성을 관리하고 독립성을 보장하여 유지보수와 확장을 용이하게 하기 위한 설계 가이드일 뿐입니다.

"함수만 잘 정리해서 선언하더라도 충분히 유지보수 가능한 코드가 된다"는 말은 어느 정도 일리가 있습니다. 하지만 함수만으로는 복잡한 시스템을 효과적으로 관리하고 확장하기 어렵습니다. 코드의 규모가 커지면 함수 간의 관계가 복잡해지고, 결국에는 유지보수가 어려워질 수 있습니다.

핵심은 유연하고 확장 가능한 코드를 만들기 위해 객체 지향의 장점을 잘 활용하는 것이지, 맹목적으로 규칙을 따르는 것이 아닙니다. 객체 지향을 제대로 이해하고 필요에 따라 적절하게 적용하는 것이 중요합니다.

결국 중요한 것은 규칙이 아니라 코드를 효율적으로 관리하고 유지보수하기 위한 최적의 방식을 선택하는 것입니다. 함수형 프로그래밍, 객체 지향 프로그래밍 등 다양한 프로그래밍 패러다임을 목적이 아니라 수단으로 활용해보면 어떨까 해요! 🙂